drivers/include/periph/adc.h

```diff
@@ -128,6 +128,25 @@ int adc_init(adc_t line);
  */
 int32_t adc_sample(adc_t line, adc_res_t res);
 
+/**
+ * @brief   Capture multiple ADC samples from one or multiple ADC lines
+ *
+ * This requires the `periph_adc_multi` feature.
+ *
+ * @experimental    This API may change to accommodate new backends and DMA
+ *                  operation.
+ *
+ * @param[in]  lines        line(s) to sample
+ * @param[in]  lines_numof  number of ADC lines to sample
+ * @param[out] bufs         target buffer(s)
+ * @param[in]  num_samples  number of samples to sample per ADC line
+ * @param[in]  res          resolution to use for conversion
+ * @param[in]  f_adc        frequency to use for conversion, leave 0 for default
+ */
+void adc_sample_multi(uint8_t lines_numof, const adc_t lines[lines_numof],
+                      size_t num_samples, uint16_t bufs[lines_numof][num_samples],
+                      adc_res_t res, uint32_t f_adc);
```

Comment:
arg, this does not work in C++, how about


```suggestion
#ifdef __cplusplus
void adc_sample_multi(uint8_t lines_numof, const adc_t lines[],
                      size_t num_samples, uint16_t **bufs,
                      adc_res_t res, uint32_t f_adc);
#else
void adc_sample_multi(uint8_t lines_numof, const adc_t lines[lines_numof],
                      size_t num_samples, uint16_t bufs[lines_numof][num_samples],
                      adc_res_t res, uint32_t f_adc);
#endif
```

Reply:
'uint16_t **bufs' and 'bufs[lines_numof][num_samples]' are very dissimilar

Reply:
@kfessel uint16_t bufs[lines_numof][numof_samples] actually does not decay into a uint16_t type but into a two dimensional array, as it should. You are correct that if you called a function with f(bufs[lines_numof][numof_samples]), the function would receive an uint16_t, but that is not the case for function declarations.

I did not find any examples that you can specify both array sizes like that in the function declaration, so I am a bit confused that the compiler does not complain about it.
However, this is a very interesting StackOverflow thread that explains the situation quite well: https://stackoverflow.com/questions/8767166/passing-a-2d-array-to-a-c-function

And it also gives a reference to a C FAQ that explicitly says that a pointer to a pointer is not the same as a two dimensional array: https://c-faq.com/aryptr/pass2dary.html
Therefore the proposed fix is not a valid solution.


The applicable solution for the issue would be approach 2 from stackoverflow with the corresponding call:
```cpp
void adc_sample_multi(uint8_t lines_numof, const adc_t lines[lines_numof],
                      size_t num_samples, uint16_t (*bufs)[lines_numof][num_samples],
                      adc_res_t res, uint32_t f_adc);

...

adc_sample_multi(..., &samples, ...);
```

HOWEVER I don't know if the C compiler allows the dynamic boundaries. Perhaps it would require const size_t lines_numof and numof_samples or even LINES_NUMOF/NUMOF_SAMPLES as preprocessor defines, so it definitively known at compile time.
The comment says that the array size has to be fixed and known at compile time, so it probably won't compile.


(On a side note: IMO it would be better to have a consistent naming like num_samples and num_lines or samples_numof and lines_numof).

Reply:
thank you @crasbe for providing more hints and looking into that. 

For me my comment was more about the *very dissimilar* and the c++ header variant not having a compatible interface to the c implementation if written that way, it was more about putting the message out fast (before any merge happens).

A simple solution might also be to hide the function declaration entirely from C++ instead of making it less kiss for C.